test/integration/apiserver/cel/validatingadmissionpolicy_test.go

```diff
@@ -900,6 +908,503 @@ func Test_ValidatingAdmissionPolicy_UpdateParamRef(t *testing.T) {
+		matchedConfigMap := &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "denied",
+				Namespace: "default",
+				Labels: map[string]string{
+					"foo": "bar",
+				},
+			},
+		}
+
+		_, err := client.CoreV1().ConfigMaps(matchedConfigMap.Namespace).Create(context.TODO(), matchedConfigMap, metav1.CreateOptions{})
+		if err == nil {
+			return false, errors.New("policy not enforced yet")
+		}
+
+		if !strings.Contains(err.Error(), "matched by object selector!") {
+			return false, err
+		}
+
+		return true, nil
+
+	}); waitErr != nil {
+		t.Errorf("timed out waiting: %v", waitErr)
+	}
+
+	allowedConfigMap := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "allowed",
+			Namespace: "default",
+		},
+	}
+
+	if _, err := client.CoreV1().ConfigMaps(allowedConfigMap.Namespace).Create(context.TODO(), allowedConfigMap, metav1.CreateOptions{}); err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+}
+
+func Test_ValidatingAdmissionPolicy_MatchByNamespaceSelector(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.CELValidatingAdmission, true)()
+	server, err := apiservertesting.StartTestServer(t, nil, []string{
+		"--enable-admission-plugins", "ValidatingAdmissionPolicy",
+	}, framework.SharedEtcd())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer server.TearDownFn()
+
+	config := server.ClientConfig
+
+	client, err := clientset.NewForConfig(config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// only configmaps in default will be allowed.
+	labelSelector := &metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      "kubernetes.io/metadata.name",
+				Operator: "NotIn",
+				Values:   []string{"default"},
+			},
+		},
+	}
+
+	policy := withValidations([]admissionregistrationv1alpha1.Validation{
+		{
+			Expression: "false",
+			Message:    "matched by namespace selector!",
+		},
+	}, withConfigMapMatch(withFailurePolicy(admissionregistrationv1alpha1.Fail, makePolicy("match-by-namespace-selector"))))
+	policy = withNamespaceSelector(labelSelector, policy)
+	_, err = client.AdmissionregistrationV1alpha1().ValidatingAdmissionPolicies().Create(context.TODO(), policy, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	policyBinding := makeBinding("match-by-namespace-selector-binding", "match-by-namespace-selector", "")
+	_, err = client.AdmissionregistrationV1alpha1().ValidatingAdmissionPolicyBindings().Create(context.TODO(), policyBinding, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	namespace := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "not-default",
+		},
+	}
+	if _, err := client.CoreV1().Namespaces().Create(context.TODO(), namespace, metav1.CreateOptions{}); err != nil {
+		t.Fatal(err)
+	}
+
+	if waitErr := wait.PollImmediate(time.Millisecond*10, wait.ForeverTestTimeout, func() (bool, error) {
```

Comment:
There seems to be a bit of overlap with this code here (the polling function in `Test_ValidatingAdmissionPolicy_MatchByObjectSelector`): https://github.com/kubernetes/kubernetes/pull/113738/commits/62e06a919a2a5bf997917e1cdda18cf1f2f4573b#diff-e1f94a3070165f52b690b0e3ca5c9920077a9ff2a9c2dfb6d73231f8b0e5b53cR952-R976

Could the common parts be extracted out maybe? It looks like the other tests also have similar polling functions.

Reply:
I felt like `wait.PollImmediate` itself is already pretty generic. The resource created inside the poll func() varies between tests though (diferent labels, namespaces, resources, etc) so I didn't feel it was worthwhile to extract it out.